Add dex score tiebreaker to initiative

jaypoulz opened the pull request:
Closes #105

jaypoulz commented:
@kakaroto Do you have a test strategy for new changes? I can test on FF & Chrome for r20, but I don't have access to Foundry VTT.

kakaroto commented:
Hey @jaypoulz, thanks for the PR and taking care of the issue #105.
I gave it a quick look and it generally looked good, other than some differences in style and `console.log` that need to be removed.
For testing strategy, I mostly test on one system (roll20 or FVTT) in Chrome and once the feature is done, I test on the other system if there was a need to write system specific changes, then I only go through a quick "let's make sure nothing is broken in Firefox" before a release.
For testing how it would be on FVTT, you can just close your Roll20 tab or click the quick settings there and set it to send rolls to the D&D Beyond dice roller, it's the same code/renderer between FVTT and the integrated Dice roller.
I'll do a quick review of the PR itself so you can fix things before you send an update and we get it merged.
Thanks again!

kakaroto commented on src/dndbeyond.pyj:
```diff
@@ -613,7 +613,9 @@ class Monster(CharacterBase):
                 modifier = ability[3]
-                sendRoll(self, "initiative", "1d20" + modifier, {"initiative": modifier})
+                score = ability[2]
+                roll_modifier = float(modifier) + float(score / 100) if self.getGlobalSetting("initiative-with-dex-score", False) else modifier
```
Due to the length of the line, please split it into :
```python
if self.getGlobalSetting("initiative-with-dex-score", False):
  score = ability[2]
  roll_modifier = float(modifier) + float(score / 100)
else:
  roll_modifier = modifier
```
to make it easier to read. I usually reserve the one-liner `x if y else z` constructs for simpler lines only.

kakaroto replied:
Seeing the later cast of the modifier into a string from the roll20 side, this reminds me that the roll modifier here should be a string, and not a float, so make sure you cast it into a string in here.

jaypoulz replied:
Do you prefer the stylistic:
```
initiative = modifier
if self.getGlobalSetting("initiative-with-dex-score", False):
  score = ability[2]
  initiative = str(float(modifier) + float(score / 100))
```

kakaroto replied:
yes, actually, I do prefer it the way you just described. Leaving optional settings to just change the behavior rather than define it. Especially since the default would be `initiative = modifier`
Good catch. Thanks!

kakaroto commented on src/dndbeyond_character.pyj:
```diff
@@ -66,12 +66,15 @@ def rollSavingThrow():
         advantage = $(".ct-combat-mobile__advantage").length > 0
-    #console.log("Initiative " + ("with" if advantage else "without") + " advantage : " + initiative)
+    console.log(character)
+    initiative = initiative if not settings.getSetting("initiative-with-dex-score", False) else initiative
```
Same as above, split the if into a non-tertiary operator.
Also, it should be `character.getGlobalSetting`, I think `settings` is only a dictionary, so `settings.getSetting` would not work.
In this case, you can also get the DEX abilities using : 
```python
   if character.getGlobalSetting("initiative-with-dex-score")
     tiebreaker = 0
     for ability in character._abilities:
        if ability[1] == "DEX":
           tiebreaker = ability[2]
```
And finally, remove the `console.log` lines

jaypoulz replied:
Would you rather me fix this in both dndbeyond.pyj & dndbeyond_character.pyj (as is the current approach) or did you want me to have dndbeyond_character.pyj call the parent class method and pass along the initiative modifier?

kakaroto replied:
The code is honestly a mess, so there isn't really any 'parent method' in this case. The 'rollInitiative' in dndbeyond.pyj is a method that exists only for the `Monster` class (the only common functions are in `CharacterBase` which is the `getSetting` and `getGlobalSetting` I think), so the change has to go in both files/classes.

jaypoulz replied:
Ah, OK. Is there any value in doing a rolling refactor such that we move this function into CharacterBase with the intention of fixing the others as you go? Or would you rather keep them the way they are so that you don't find yourself in a 1/2 way refactored state?

kakaroto replied:
A refactor at this point would be more of a rewrite. Statblocks and character sheets are also very different in D&D Beyond. A roll Initiative for a monster is about simply adding dexterity, but also modifying the statblock to add my own 'Roll Initiative' button for convenience.
For a character, it's about checking the initiative value in the sidebar, verifying if there are 'advantage on initiative' modifiers due to class features, etc.. 
The way the two work is too different to have anything merged. If you had a common function name for rolling initiative, you'd just have them override it both, so it's not worth pursuing at this time.

Note: Beyond20 was initially written as a thought experiment, a "would it be possible to... ?" question, from before I knew how to write javascript (explains the .pyj) and from before I knew html/css (I still don't). It worked so easily and quickly, and got so much traction that it quickly evolved, but the code is unfortunately still very messy. I apologize for that :)

jaypoulz replied:
I think you've described the origin story for just about any project that's ever come to be. :)

No need for apologies - it works! That's a testament to itself!

Rewrites can be very good and very bad. If you actually have the time (and energy) you need to follow through, then it could become the new base and you could incorporate all you learned. However, its easier to iterate on working code then to hack on vaporware. Also, the longer you wait to rewrite, the more you'll have to write to achieve feature parity. Well-written software can often lose to worse written software that was good enough to work when people needed it.

You're the maintainer, so I obviously want to defer to your vision.

It might make sense to move towards a Model-View-Controller type organization. The abstraction I would make would be that characters and monsters are special cases of a creature, similar to how the game is structured. Creatures (models) would have abilities, speed, type, size, etc and corresponding (controller) methods to interact with them (e.g. rollInitiative). Your "views" would be the code that inject the buttons into the page. You would also need controllers to parse the dndbeyond page specifics.

I don't think the underlying tech you pick matters all that much once you have a portable/modular architecture.

jaypoulz commented:
Awesome - I'll clean up the PR once I've implemented the character sheet based initiative roll, test it, and ping you for review when its ready. Thanks!

kakaroto commented on src/roll20.pyj:
```diff
@@ -49,7 +49,8 @@ def escapeRoll20Macro(text):
     roll = "[[" + dice
     for m in modifiers:
-        mod = modifiers[m]
+        mod = str(modifiers[m])
```
shouldn't be necessary to cast it to a string as it should always be sent as a string from the dndbeyond side

kakaroto commented on src/roll20.pyj:
```diff
@@ -282,6 +284,7 @@ def rollInitiative(request, custom_roll_dice=""):
             dice = ROLL20_INITIATIVE_ADVANTAGE_QUERY
         else:
             dice = "1d20"
+        console.log(request.initiative)
```
remember the cleanup the `console.log` lines before final PR

kakaroto commented on src/settings.pyj:
```diff
@@ -85,6 +85,11 @@ options_list = {
                            "type": "bool",
                            "default": True},
 
+    "initiative-with-dex-score": {"title": "Add dexterity score tiebreaker to initiative roll.",
```
I would rename the setting into `initiative-tiebreaker` rather than `initiative-with-dex-score` (what it does, rather than how it does it) and adjust the title accordingly. I'm not super good with naming stuff, but my suggestion would be : 
- title: Add tie breaker to initiative rolls.
- description: Adds the dexterity score as a decimal to initiative rolls to break any initiative ties.

jaypoulz commented:
@kakaroto ready for review!
Tested for on roll20 and using no-VTT renderer on Chrome.
Still need to sanity check ffox.

jaypoulz commented:
Also I didn't add the compiled js since my rapydscrypt seems to compile differently. One suggestion would be to build a release branch and only keep the compiled js in that. This way you can keep master clean for development, and you can still have the compiled js in the downloadable releases.

kakaroto commented on src/dndbeyond_dice.pyj:
```diff
@@ -188,6 +188,7 @@ class DNDBRoll(Beyond20BaseRoll):
                 self._total += part.roll()
             else:
                 self._total += part
+        self._total = Math.round((self._total + Number.EPSILON) * 100) / 100
```
Why are you adding epsilon here?

jaypoulz replied:
I actually made a note of this in the commit summary.
This is a rounding trick to ensure that edge cases are rounded correctly.
https://stackoverflow.com/questions/11832914/round-to-at-most-2-decimal-places-only-if-necessary

It's probably not necessary in this case.

kakaroto commented on src/dndbeyond.pyj:
```diff
@@ -613,9 +613,14 @@ class Monster(CharacterBase):
+                initiative = modifier
+                if self.getGlobalSetting("initiative-tiebreaker", False):
+                    score = ability[2]
+                    initiative= str(float(modifier) + float(score) / 100)
```
Space missing between `initiative` and `=`

kakaroto commented on src/dndbeyond_character.pyj:
```diff
@@ -66,12 +66,18 @@ def rollSavingThrow():
     #console.log("Initiative " + ("with" if advantage else "without") + " advantage : " + initiative)
+
+    if character.getGlobalSetting("initiative-tiebreaker", False):
+       dexterity_score = [x[2] for x in character._abilities if x[1] == 'DEX'][0]
```
This will crash if there is no page DEX ability, which could happen if user loads their sheet on a tab other than the abilities (not entirely sure, but i remember there were cases when it wouldn't find abilities, that's why it always checks elsewhere uf it finds it in the abilities array)

jaypoulz commented:
I can split the ws fix with the other commit - that's just me being lazy since my emacs is set to autostrip white space. Splitting the commit will take me just a moment.

As for the missing dex score problem - is there an alternative way of getting the dex score if it's missing from _abilities? If not, this sounds like I should just pass initiative as is if I encounter this situation.

kakaroto commented:
No problem. I use `git add -p` for whenever I want to commit something, allows me to review my stuff as well. But like I said, it's not a big deal since the changes are small.
For the dex, it was in my original comment : https://github.com/kakaroto/Beyond20/pull/107#discussion_r383442418
```python
 if character.getGlobalSetting("initiative-with-dex-score")
     tiebreaker = 0
     for ability in character._abilities:
        if ability[1] == "DEX":
           tiebreaker = ability[2]
```
That way if DEX isn't found, the tiebreaker is 0, and use that variable for the math.

I just saw what `Number.EPSILON` is (I thought it was Euler's number 2.71828), so I guess it makes no difference to the actual value to have it added, but I still don't understand the need to do it, along with the `* 100 / 100`.
If it's because JS causes the number to be `10.1200000000001` for example, instead of `10.12`, then how about doing instead : `parseFloat(self._total.toFixed(2))` with a comment above it to explain why it needs to be there?

kakaroto commented:
Ah, I just saw the commit log message that links to https://stackoverflow.com/questions/11832914/round-to-at-most-2-decimal-places-only-if-necessary
 (sorry for missing it yesterday, I did the review on my phone while I was out).
That makes sense, and it's good, it can stay as is, but I would put the stackoverflow link in a comment above that line, as it would be easier for the future to figure out when reading the code rather than trying to find the the link from the commit which added the line.
Thanks!

kakaroto commented:
Thanks for all the changes. Nicely done. I like the `reduce` solution for the ability modifier.
I've merged it now, thanks! I hope to see more from you in the future :)

jaypoulz commented:
Ready for final review!

- Separated the commits by whitespace & content.
- Removed the epsilon bit since it's covering edge cases we aren't going to hit
- Added some logic to ensure we render the same with and without the flag set.
- Tested {-x,0,+x} for characters & monsters are both d&d beyond and the beyond20/vtt renderer {Chrome, FFox}

jaypoulz commented:
LOL you beat me to it. :) Thanks!

kakaroto commented:
Yeah, I saw it pop in the discord #bots channel, so I did the quick review and merge :)
